### ISSUE TITLE
Object overriding array in models silently on update

### ISSUE THREAD

**Issue by wesleimarinho** [Automattic/mongoose#6532]
**Do you want to request a *feature* or report a *bug*?**
bug

**What is the current behavior?**
I expected to mongoose throw an error when updating an array field with an object value, but instead it saves the object silently.

**If the current behavior is a bug, please provide the steps to reproduce.**

```javascript
#!/usr/bin/env node
'use strict';

const mongoose = require('mongoose');
mongoose.connect('mongodb://localhost/test');
const conn = mongoose.connection;
const Schema = mongoose.Schema;

const sub = new Schema({
  x: String
});

const schema = new Schema({
  name: String,
  arr: [sub]
});

const Test = mongoose.model('test', schema);

const test = {name: "Xyz"};
const arr1 = [{x: "Z"}];
const obj1 = {x: "Y"};

async function run() {
  await conn.dropDatabase();
  test.arr = arr1;
  await Test.create(test);
  
  await Test.update({name: "Xyz"}, {$set: {"arr": obj1}});
  
  return conn.close();
}

run().catch(console.error);

```

**What is the expected behavior?**
To throw an error.

**Please mention your node.js, mongoose and MongoDB version.**
Mongoose 5.1.12
NodeJS: 8.9.4
MongoDB: 3.6.3

**Comment by lineus** [comment]
@wesleimarinho When you inspect the document in the collection after running the update command in your example you'll see that mongoose doesn't just save the object as it is in the update doc. 

Mongoose casts it to the schemaType as defined in the schema, in this case an array of embedded docs. 

If you add a findOne and console.log the doc from your example:
```
  let found = await Test.findOne({});
  console.log(found);
```

the output is:
```
issues: ./6532.js
{ _id: 5b0c95a90f75430970ee42a6,
  name: 'Xyz',
  arr: [ { _id: 5b0c95aa0f75430970ee42a8, x: 'Y' } ],
  __v: 0 }
issues:
```
arr now contains one element, the doc from the update command cast according the schema.

**Comment by wesleimarinho** [comment]
Thanks @lineus 

I understand the conversion, however, when I open the database in Compass it shows an Object saved in the property and not an Array. 

Shouldn't be the case to convert as Array prior saving as defined in the model, or issue an warning at least?

**Comment by wesleimarinho** [comment]
Below are some screenshots from Compass:

Document after `create()`:

![image](https://user-images.githubusercontent.com/30837062/40659735-82669230-6325-11e8-8d7e-0709c67a86fd.png)

Document after `update()`:

![image](https://user-images.githubusercontent.com/30837062/40659784-a2d7f68a-6325-11e8-8703-b131161f06ef.png)

**Comment by lineus** [comment]
@wesleimarinho you're right, using the mongo shell to inspect the doc I see:
```
issues: mongo --quiet test
> db.tests.find()
{ "_id" : ObjectId("5b0d4c4417c0ec2af97e1de3"), "name" : "Xyz", "arr" : { "_id" : ObjectId("5b0d4c4517c0ec2af97e1de5"), "x" : "Y" }, "__v" : 0 }
>
```
so it's the findOne that's casting the value from the db correctly, but not update.

I'll dig into this further.

**Comment by wesleimarinho** [comment]
You are right, the problem is with `update`. I made a new test with create and the Object was cast to Array prior saving correctly.

**Comment by lineus** [comment]
@wesleimarinho The fix for this will be in 5.1.4 which should be published by early next week.
